pkg/host/internal/sriov/sriov.go

```diff
@@ -823,32 +894,98 @@ func (s *sriov) encapTypeToLinkType(encapType string) string {
 
 // create required udev rules for PF:
 // * rule to disable NetworkManager for VFs - for all modes
+// * rule to rename VF representors - only for switchdev mode
 func (s *sriov) addUdevRules(iface *sriovnetworkv1.Interface) error {
 	log.Log.V(2).Info("addUdevRules(): add udev rules for device",
 		"device", iface.PciAddress)
-	// TODO add creation of switchdev-related UDEV rules
-	return s.udevHelper.AddUdevRule(iface.PciAddress)
+	if err := s.udevHelper.AddUdevRule(iface.PciAddress); err != nil {
+		return err
+	}
+	if sriovnetworkv1.GetEswitchModeFromSpec(iface) == sriovnetworkv1.ESwithModeSwitchDev {
+		portName, err := s.networkHelper.GetPhysPortName(iface.Name)
+		if err != nil {
+			return err
+		}
+		switchID, err := s.networkHelper.GetPhysSwitchID(iface.Name)
+		if err != nil {
+			return err
+		}
+		if err := s.udevHelper.AddVfRepresentorUdevRule(iface.PciAddress, iface.Name, switchID, portName); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
 // remove all udev rules for PF created by the operator
 func (s *sriov) removeUdevRules(pciAddress string) error {
 	log.Log.V(2).Info("removeUdevRules(): remove udev rules for device",
 		"device", pciAddress)
-	// TODO add support for removing switchdev-related UDEV rules
-	return s.udevHelper.RemoveUdevRule(pciAddress)
+	if err := s.udevHelper.RemoveUdevRule(pciAddress); err != nil {
+		return err
+	}
+	return s.udevHelper.RemoveVfRepresentorUdevRule(pciAddress)
 }
 
 // create VFs on the PF
 func (s *sriov) createVFs(iface *sriovnetworkv1.Interface) error {
+	expectedEswitchMode := sriovnetworkv1.GetEswitchModeFromSpec(iface)
 	log.Log.V(2).Info("createVFs(): configure VFs for device",
-		"device", iface.PciAddress, "count", iface.NumVfs)
-	if iface.NumVfs == s.dputilsLib.GetVFconfigured(iface.PciAddress) {
+		"device", iface.PciAddress, "count", iface.NumVfs, "mode", expectedEswitchMode)
+
+	if iface.NumVfs == s.dputilsLib.GetVFconfigured(iface.PciAddress) && expectedEswitchMode == s.getCurrentEswitchMode(iface.PciAddress) {
 		log.Log.V(2).Info("createVFs(): device is already configured",
-			"device", iface.PciAddress, "count", iface.NumVfs)
+			"device", iface.PciAddress, "count", iface.NumVfs, "mode", expectedEswitchMode)
 		return nil
 	}
-	// TODO add support for VF creation in switchdev mode
-	return s.SetSriovNumVfs(iface.PciAddress, iface.NumVfs)
+	return s.setEswitchModeAndNumVFs(iface.PciAddress, expectedEswitchMode, iface.NumVfs)
+}
+
+func (s *sriov) getCurrentEswitchMode(pciAddr string) string {
+	log.Log.V(2).Info("getCurrentEswitchMode(): get eswitch mode", "device", pciAddr)
+	mode, err := s.GetNicSriovMode(pciAddr)
```

Comment:
you already assume an empty eswitch mode in spec and status means Legacy.

perhaps modify `GetNicSriovMode()` to return "legacy" instead of an empty string in case device does not support devlink.

then, we either fail or return the mode. WDYT ?

Reply:
I applied this change